graylog2-server/src/main/java/org/graylog2/cluster/NodeServiceImpl.java

```diff
@@ -153,24 +147,38 @@ public void markAsAlive(NodeId node, boolean isLeader, URI restTransportAddress)
         }
     }
 
+    private Map<String, Object> recentHeartbeat() {
+        return Map.of("$where", "this.last_seen >= Timestamp(new Date().getTime() / 1000 - " + pingTimeout + ", 1)");
+    }
+
     @Override
     public boolean isOnlyLeader(NodeId nodeId) {
-        BasicDBObject query = new BasicDBObject();
-        query.put("type", Node.Type.SERVER.toString());
-        query.put("last_seen", new BasicDBObject("$gte", Tools.getUTCTimestamp() - pingTimeout));
-        query.put("node_id", new BasicDBObject("$ne", nodeId.getNodeId()));
-        query.put("is_leader", true);
-
-        return query(NodeImpl.class, query).size() == 0;
+        final BasicDBObject query = new BasicDBObject(Map.of(
+                "$and", List.of(
+                        recentHeartbeat(),
+                        Map.of(
+                                "type", Node.Type.SERVER.toString(),
+                                "node_id", new BasicDBObject("$ne", nodeId.getNodeId()),
+                                "is_leader", true
+                        )
+                )
+        ));
+
+        return query(NodeImpl.class, query).isEmpty();
     }
 
     @Override
     public boolean isAnyLeaderPresent() {
-        BasicDBObject query = new BasicDBObject();
-        query.put("type", Node.Type.SERVER.toString());
-        query.put("last_seen", new BasicDBObject("$gte", Tools.getUTCTimestamp() - pingTimeout));
-        query.put("is_leader", true);
-
-        return query(NodeImpl.class, query).size() > 0;
+        final BasicDBObject query = new BasicDBObject(Map.of(
+                "$and", List.of(
+                        recentHeartbeat(),
+                        Map.of(
+                                "type", Node.Type.SERVER.toString(),
+                                "is_leader", true
+                        )
+                )
+        ));
+
+        return !query(NodeImpl.class, query).isEmpty();
```

Comment:
If we only care it the result is empty or not, maybe we should replace `query()` with `count()`, to not fetch unnecessary data?

Reply:
:+1: